aptos-move/framework/aptos-framework/releases/artifacts/current/build/AptosFramework/sources/Table.move

```diff
@@ -0,0 +1,131 @@
+/// This module provides a temporary solution for tables by providing a layer on top of Vector
+module AptosFramework::Table {
+    use Std::Errors;
+    use Std::Option;
+    use Std::Vector;
+
+    const EKEY_ALREADY_EXISTS: u64 = 0;
+    const EKEY_NOT_FOUND: u64 = 1;
+
+    struct Table<Key: drop + store, Value: drop + store> has drop, store {
+      data: vector<TableElement<Key, Value>>,
+    }
+
+    struct TableElement<Key: drop + store, Value: drop + store> has drop, store {
+        key: Key,
+        value: Value,
+    }
+
+    public fun create<Key: drop + store, Value: drop + store>(): Table<Key, Value> {
+        Table {
+            data: Vector::empty(),
+        }
+    }
+
+    public fun borrow<Key: drop + store, Value: drop + store>(
+        table: &Table<Key, Value>,
+        key: &Key,
+    ): &Value {
+        let maybe_idx = find(table, key);
+        assert!(Option::is_some(&maybe_idx), Errors::invalid_argument(EKEY_NOT_FOUND));
+        let idx = *Option::borrow(&maybe_idx);
+        &Vector::borrow(&table.data, idx).value
+    }
+
+    public fun borrow_mut<Key: drop + store, Value: drop + store>(
+        table: &mut Table<Key, Value>,
+        key: &Key,
+    ): &mut Value {
+        let maybe_idx = find(table, key);
+        assert!(Option::is_some(&maybe_idx), Errors::invalid_argument(EKEY_NOT_FOUND));
+        let idx = *Option::borrow(&maybe_idx);
+        &mut Vector::borrow_mut(&mut table.data, idx).value
+    }
+
+    public fun contains_key<Key: drop + store, Value: drop + store>(
+        table: &Table<Key, Value>,
+        key: &Key,
+    ): bool {
+        Option::is_some(&find(table, key))
+    }
+
+    public fun insert<Key: drop + store, Value: drop + store>(
+        table: &mut Table<Key, Value>,
+        key: Key,
+        value: Value,
+    ) {
+        let maybe_idx = find(table, &key);
+        assert!(Option::is_none(&maybe_idx), Errors::invalid_argument(EKEY_ALREADY_EXISTS));
```

Comment:
it is a fair question! and I probably should move toward that direction ;)

Reply:
I'll say this much, the ergonomics of assuming this makes using the table a lot easier. I suspect in most cases folks know whether or not there is an entry. If there is an entry, you will need to handle it in a special way. If there isn't an entry, then you don't want to deal with anything special. More explicitly:

* If we return an Option<(...)>, the option is not be a tuple, it can be a value only
* If we return an Option, even if it is a none, it must be handled -- i.e., destroyed
* If there was already a value, we probably needed to do some sort of merge and update operation (e.g., we're doing a borrow_mut call on the table instead of an insert)

I think the reality is that Tables in a smart contract are very different than caches and so the semantics change.

I actually did the work of converting this over and it felt really unintuitive as it added a lot more boilerplate code.